groups/bsl/bslstl/bslstl_monostate.t.cpp

```diff
@@ -0,0 +1,309 @@
+#define LOOP3_ASSERT BSLS_BSLTESTUTIL_LOOP3_ASSERT
+#define LOOP4_ASSERT BSLS_BSLTESTUTIL_LOOP4_ASSERT
+#define LOOP5_ASSERT BSLS_BSLTESTUTIL_LOOP5_ASSERT
+#define LOOP6_ASSERT BSLS_BSLTESTUTIL_LOOP6_ASSERT
+
+#define Q BSLS_BSLTESTUTIL_Q    // Quote identifier literally.
+#define P BSLS_BSLTESTUTIL_P    // Print identifier and value.
+#define P_ BSLS_BSLTESTUTIL_P_  // P(X) without '\n'.
+#define T_ BSLS_BSLTESTUTIL_T_  // Print a tab (w/o newline).
+#define L_ BSLS_BSLTESTUTIL_L_  // current Line number
+
+#define RUN_EACH_TYPE BSLTF_TEMPLATETESTFACILITY_RUN_EACH_TYPE
+
+// ============================================================================
+//                  NEGATIVE-TEST MACRO ABBREVIATIONS
+// ----------------------------------------------------------------------------
+
+#define ASSERT_SAFE_PASS(EXPR) BSLS_ASSERTTEST_ASSERT_SAFE_PASS(EXPR)
+#define ASSERT_SAFE_FAIL(EXPR) BSLS_ASSERTTEST_ASSERT_SAFE_FAIL(EXPR)
+#define ASSERT_PASS(EXPR) BSLS_ASSERTTEST_ASSERT_PASS(EXPR)
+#define ASSERT_FAIL(EXPR) BSLS_ASSERTTEST_ASSERT_FAIL(EXPR)
+#define ASSERT_OPT_PASS(EXPR) BSLS_ASSERTTEST_ASSERT_OPT_PASS(EXPR)
+#define ASSERT_OPT_FAIL(EXPR) BSLS_ASSERTTEST_ASSERT_OPT_FAIL(EXPR)
+
+//=============================================================================
+//                         GLOBAL FUNCTIONS FOR TESTING
+//-----------------------------------------------------------------------------
+
+//=============================================================================
+//                             USAGE EXAMPLE
+//-----------------------------------------------------------------------------
+
+//=============================================================================
+//                              MAIN PROGRAM
+//-----------------------------------------------------------------------------
+
+int main(int argc, char *argv[])
+{
+    int  test    = argc > 1 ? atoi(argv[1]) : 0;
+    bool verbose = argc > 2;
+    //  bool veryVerbose         = argc > 3;
+    //  bool veryVeryVerbose     = argc > 4;
+    //  bool veryVeryVeryVerbose = argc > 5;
+
+    printf("TEST " __FILE__ " CASE %d\n", test);
+
+    switch (test) {
+      case 0:
+
+      case 5: {
+        // --------------------------------------------------------------------
+        // TESTING 'hashAppend'
+        //
+        // Concerns:
+        //: 1 Hashing a monostate object is equivalent to appending 'false'
+        //:   to the hash.
+        //
+        // Plan:
+        //: 1 Create a monostate object and verify that hashing it yields the
+        //:   same value as hashing 'true'. [C-1]
+        //
+        // Testing:
+        //   void hashAppend(HASHALG& hashAlg, const monostate &);
+        // --------------------------------------------------------------------
+
+        if (verbose)
+            printf("\nTESTING 'hashAppend'"
+                   "\n====================\n");
+
+        {
+            bsl::monostate obj;
```

Comment:
Is this suggestion related to some sort of test writing policy ?

Reply:
No, it's just based on the general idea that if we make objects `const` then we can check that calls that are supposed to accept `const` objects do in fact accept `const` objects. However, now that I look at this again, I realize that `monostate` is not one of the classes you gave a user-declared default constructor to allow this to work in old compilers, so I guess my suggestion doesn't work and it's fine to ignore it because I know you wouldn't screw up something like this anyway.

Reply:
How did we get to discussing user-declared default constructors for old compilers ? I'm missing a step in your thought process. 

I don't mind making all these test objects const, I just wanted to understand the request. Made this one and all the other ones in this test suite const.

Btw, I noticed a few unused variables warning so have sprinkled a usage in the tests that required it to stifle the warning.

Reply:
Ah. That's a bit of a mess.
The reason in_place_t has a constructor at all is to make it explicit, rather than to make it const-default-constructible (the explicit default constructor is called out in the standard). All these tag types have no members, so they are all const-default-constructible - or at least they are after the resolution of CWG253. I checked the oldest gcc and clang compiler on wandbox in  C++03 mode and it doesn't look like compilers ever detected a const initialisation of a class type without a default user-provided default constructor anyway (which is why me adding const to these tests didn't trigger any errors - at least on my machine).

The reason monostate doesn't have any constructors is because the standard doesn't give it any, so neither did I. Theoretically, it is now const-default-constructible, however, also theoretically, there could be some older compilers that might complain. I suggest I make these test monostate objects const. If there are old compilers that we use that do error before CWG253, at least we'll see it. We could also add a user defined default constructor before C++17, but it seems unlikely to be needed, and we do lose out on having this type be a POD.

Reply:
Fair enough

Reply:
If it doesn't work on the AIX/Sun compilers, we'll see it during the matrix build and the fix will be easy.